functions/grapher/[slug].ts

```diff
@@ -5,17 +5,30 @@ export const onRequestGet: PagesFunction = async (context) => {
 
     const { request, env, params } = context
 
-    const slug = params.slug as string
+    const originalSlug = params.slug as string
     const url = new URL(request.url)
 
     // All our grapher slugs are lowercase by convention.
     // To allow incoming links that may contain uppercase characters to work, we redirect to the lowercase version.
+    // We also then resolve redirects using the /grapher/_grapherRedirects.json file.
     // We do this before we even know that the page exists, so even if we redirect it might still be a 404, or go into another redirect.
-    if (
-        url.pathname !== url.pathname.toLowerCase() &&
-        url.pathname !== "/grapher/embedCharts.js"
-    ) {
-        const redirUrl = url.pathname.toLowerCase() + url.search
+    let currentSlug = originalSlug.toLowerCase()
+    const redirects = await env.ASSETS.fetch(
+        new URL("/grapher/_grapherRedirects.json", url),
+        { cf: { cacheTtl: 2 * 60 } }
+    )
+        .then((r): Promise<Record<string, string>> => r.json())
+        .catch((e) => {
+            console.error("Error fetching redirects", e)
```

Comment:
I am worried this will go unnoticed. Would it be possible to send the error to bugsnag? (if yes, I'll reuse it in other functions)

Reply:
Ah yes, good call probably.
But I just checked, and Bugsnag doesn't seem to be compatible with CF Workers currently: https://github.com/bugsnag/bugsnag-js/issues/637 - it either expects some browser APIs or some node APIs to be present.

Overall, I agree that we need better error handling & error tracking in our various workers & functions.

Reply:
Thanks for checking!